### DISCUSSION TITLE
M2O relationship permissions to avoid "claiming someone else's item"

### DISCUSSION THREAD

**Discussion by AKockWorth** [directus/directus#10733] [Archive]
I have two collections with a many-to-one relationship between them. It's really meant as a one-to-one relationship, but what matters is that the many side has the foreign key. For sake of example let's say it's the following `User has one Profile` and that both sides have a field pointing to the other collection.

I have set permissions on `Profile` so that only the profile of the user can be read by that user. This works:

```json
{ "User": { "Id": { "_eq": "$CURRENT_USER" } } }
```

However, I also want avoid the user being able to CLAIM someone else's profile if they do an update through the API. So if they do a `PATCH` on `User` with the following payload:

```
{
 "Id": 1,
 "Profile": { "Id": 5, ... },
 ...
}
```

Where the profile with ID 5 belongs to a different user.

If I use the same filter rule as I do for the READ permissions as field validation in an update permission rule, I get an error saying that `"User" must be of type object`.

The goal is to have Profile be nullable and to allow _new_ Profiles to be created. But once created, the user is no longer allowed to change the relationship to point to a different pre-existing `Profile`. How do I achieve this?

**Comment by rijkvanzanten** [comment]
> The goal is to have Profile be nullable and to allow new Profiles to be created. But once created, the user is no longer allowed to change the relationship to point to a different pre-existing Profile. How do I achieve this?

There's two ways in the API the user can associate the existing account, either through:

```json
{ 
  "Profile": {
    "Id": 5
  }
}
```

or

```json
{
  "Profile": 5
}
```

so if you setup the permission to require an object with a field Id, the second is no longer valid. You can get around this by using an OR filter to allow both these flavors

**Reply by AKockWorth** [reply]
Ah, that explains the error message, I guess. But how do I make it impossible to  _change_ the `Profile` id when updating the `User` item?

Since the update operations supports transitive/nested items for related collections, it isn't clear to me whether I need to be setting a permission (or perhaps a field validation) on the `User` or the `Profile` collection. Profile makes the most sense, but since I'm updating _through_ the user, perhaps it doesn't work that way...

**Reply by rijkvanzanten** [reply]
You're correct. The nested object is simply used to create a new update operation on the related collection, in this case Profile. At that point, the permissions for Profile control what can be changed (if anything).

**Comment by lgvld** [comment]
Still an open question, right?

When there are two collections where the many side contains the foreign key, i.e. we have a `Message` collection and an `Attachments` collection where each attachment points to a message.

Let’s say everyone can create both messages and attachments. When someone creates an attachments, he/she specifies a message identifier to satisfy the relationship. Nothing weird so far, right?

What can be done to prevent someone (when creating an attachment) to specify the identifier of a message he/she doesn’t have access to (notably if he/she didn’t created the message in the first place)?

When setting rules using the roles/permissions system, it requires an object (in order to access its fields), i.e. we cannot simply create an attachment by specifying `"message": "12345"` as the rule verifying if `message.user_created.id` equals `$CURRENT_USER.id` won’t apply. I was actually expecting Directus to dynamically load the object we are refering to (when only passing an identifier), allowing rules to have access to its fields.

Any insight to share?

Thank you very much.

Regards.

**Reply by rijkvanzanten** [reply]
That's a very good question. There's currently no way to dynamically limit the items you're allowed to select in the permissions layer. The only thing you could do is hardcode the IDs per role, but that's obviously not workable in a system where the allowed related items constantly changes (through the use of the `_in` operator). 

It sounds like you'd be looking for some way to use a sub-query in the permissions configuration, so you can set something like:

```
message
  _in
    (messages.id where messages.user_created = $CURRENT_USER)
```

That'd be an amazing new powerful filter operator type to add!

**Reply by lgvld** [reply]
Yes, exactly the idea!

Thank you very much for your answer, Rijk. And thank you for your incredible work on Directus!

Sadly I discarded Directus from my project & ended up writing my own complete API as I fear this precise problem could lead to real trouble later on (and it won’t be easy to fix).

Regards.

**Comment by samdze** [comment]
I think being able to define more complex and dynamic permissions is something that's needed in almost any project that goes beyond the simplicity of a blog, for example.
A SaaS would require much more customization and powerful permission features.

Something like @rijkvanzanten sub-queries in permissions configuration would be **very** welcome.
The more we can come close to PostgreSQL RLS policies the better.

E.g. I was trying to come up with a schema like this:
**profiles** (user data, O2O with directus_users and a M2M with projects, using profiles_projects)
**projects** (project data, with a M2M relationship with profiles, using profiles_projects)
**profiles_projects** (role of the user inside the project, junction table)

I want users to be able to read their profile data but also the profile data of every user that has at least one project in common with them.
I didn't find a way to make this work, at the moment.
EDIT: MAYBE I found a way, still, I think the "_some"/"_none" functionality together with the UI and the documentation should be improved.

Also, didn't find a way to expose fields conditionally based on dynamic rules, right now I can only choose which fields to expose based on the role.
